library/red-black/map.lisp

```diff
@@ -0,0 +1,172 @@
+                                                                                           (Optional :value)))))
+  (define (replace-or-insert mp k v)
+    "Update MP to associate K with V.
+
+If MP already contains a mapping for K, replace it and return the old value."
+    (let (%Map tre) = mp)
+    (let (Tuple new-tre removed-pair) = (tree:replace-or-insert tre (MapPair k v)))
+    (Tuple (%Map new-tre)
+           (coalton-library/classes:map value removed-pair)))
+
+  (declare insert-or-replace ((Ord :key) => ((Map :key :value) -> :key -> :value -> (Map :key :value))))
+  (define (insert-or-replace mp k v)
+    "Update MP to associate K with V.
+
+If MP already contains a mapping for K, replace it and discard the old value.
+
+Like `replace-or-insert', but prioritizing insertion as a use case."
+    (let (%Map tre) = mp)
+    (%Map (tree:insert-or-replace tre (MapPair k v))))
+
+  (declare remove ((Ord :key) => ((Map :key :value) -> :key -> (Optional (Map :key :value)))))
+  (define (remove mp k)
+    "Remove the mapping associated with K in MP. If K does not have a value in MP, return None."
+    (coalton-library/classes:map %Map
+                                 (match mp
+                                   ((%Map tre) (tree:remove tre (JustKey k))))))
+
+  (declare entries ((Map :key :value) -> (iter:Iterator (Tuple :key :value))))
+  (define (entries mp)
+    "Iterate over the (key value) pairs in MP, sorted by the keys in least-to-greatest order."
+    (match mp
+      ((%Map tre) (coalton-library/classes:map into (tree:increasing-order tre)))))
+
+  (declare keys ((Map :key :value) -> (iter:Iterator :key)))
+  (define (keys mp)
+    "Iterate over the keys in MP, sorted least-to-greatest."
+    (match mp
+      ((%Map tre) (coalton-library/classes:map key (tree:increasing-order tre)))))
+
+  (declare values ((Map :key :value) -> (iter:Iterator :value)))
+  (define (values mp)
+    "Iterate over the values in MP, sorted by their corresponding keys in least-to-greatest order."
+    (match mp
+      ((%Map tre) (coalton-library/classes:map value (tree:increasing-order tre)))))
+
+  (define-instance ((Eq :key) (Eq :value) => (Eq (Map :key :value)))
+    (define (== left right)
+      (iter:elementwise-match! == (entries left) (entries right))))
+
+  (define-instance ((Hash :key) (Hash :value) => (Hash (Map :key :value)))
+    (define (hash mp)
+      (iter:elementwise-hash! (entries mp))))
+
+  (declare collect! ((Ord :key) => ((iter:Iterator (Tuple :key :value)) -> (Map :key :value))))
+  (define (collect! iter)
+    "Construct a Map containing all the (key value) pairs in ITER.
+
+If ITER contains duplicate keys, later values will overwrite earlier values."
+    (iter:fold! (fn (mp tpl)
+                  (uncurry (insert-or-replace mp) tpl))
+                empty
+                iter))
+
+  (declare update ((Ord :key) => (:value -> :value) -> (Map :key :value) -> :key -> (Optional (Map :key :value))))
```

Comment:
Could this map `(:value1 -> :value2)`?

Reply:
`update` applies the function to only a single value in the map; it is not `fmap`. `replace` is `update` on `const`.